Commit without pre-commit hooks action is independent on prefix

kschweiger opened the pull request:
As discussed in #1984, this PR modifies the code such that the `w` command always commits without pre-commit hooks. The `skipHookPrefix` is still considered when starting the commit messaged with the prefix in the regular commit command using `c`.

Skipping pre-commits also works when switching to the editor and adding a co author from the the `w` command

I think this PR could also remove the `SkipHookPrefixNotConfigured` text in english and all internationalisations but I was not sure how this is handled. 

* [x] Cheatsheets are up-to-date (run `go generate ./...`)
* [x] Code has been formatted (see [here](https://github.com/jesseduffield/lazygit/blob/master/CONTRIBUTING.md#code-formatting))
* [x] Tests have been added/updated (see [here](https://github.com/jesseduffield/lazygit/blob/master/pkg/integration/README.md) for the integration test guide)
* [ ]  ~~Text is internationalised (see [here](https://github.com/jesseduffield/lazygit/blob/master/CONTRIBUTING.md#internationalisation))~~
* [ ]  ~~If a new UserConfig entry was added, make sure it can be hot-reloaded (see [here](https://github.com/jesseduffield/lazygit/blob/master/docs/dev/Codebase_Guide.md#using-userconfig))~~
* [ ] ~~Docs have been updated if necessary~~
* [x] You've read through your own file changes for silly mistakes etc

<!--
Be sure to name your PR with an imperative e.g. 'Add worktrees view'
see https://github.com/jesseduffield/lazygit/releases/tag/v0.40.0 for examples
-->

stefanhaller commented on pkg/commands/git_commands/commit_test.go:
```diff
@@ -63,20 +64,23 @@ func TestCommitCommitCmdObj(t *testing.T) {
 		{
 			testName:             "Commit with --no-verify flag",
 			summary:              "WIP: test",
+			verify:               true,
```
If we do what I suggest above (keeping the prefix logic in `CommitCmdObj`), then it would be good to add two more test cases here for the different combinations of prefix vs flag.

kschweiger replied:
See [97e05e5](https://github.com/jesseduffield/lazygit/pull/4374/commits/97e05e5febcf61b298c60ab156ccf7beb1fd2a6f)

stefanhaller commented on pkg/commands/git_commands/patch.go:
```diff
@@ -303,7 +303,7 @@ func (self *PatchCommands) PullPatchIntoNewCommit(
 	}
 
-	if err := self.commit.CommitCmdObj(commitSummary, commitDescription).Run(); err != nil {
+	if err := self.commit.CommitCmdObj(commitSummary, commitDescription, false).Run(); err != nil {
```
This is a regression; previously it was possible to move a custom patch into a new commit with a WIP prefix and have this skip the hooks; now this is no longer possible. Keeping the prefix logic in CommitCmdObj would fix this.

Note that there are other problems while working with custom patches currently doesn't work well with WIP prefixes; see #3532. But those are unrelated to your PR and can be fixed separately.

stefanhaller commented on pkg/gui/controllers/helpers/merge_and_rebase_helper.go:
```diff
@@ -481,7 +481,7 @@ func (self *MergeAndRebaseHelper) SquashMergeCommitted(refName, checkedOutBranch
 			"currentBranch": checkedOutBranchName,
 		})
-		err = self.c.Git().Commit.CommitCmdObj(message, "").Run()
+		err = self.c.Git().Commit.CommitCmdObj(message, "", false).Run()
```
Another regression: when setting the squashMergeMessage config to start with the skipHooksPrefix, then we would previously skip the hooks when committing the result of a squash merge. You might argue whether this was intentional, but I do think it makes sense. Again, keeping the prefix logic in CommitCmdObj fixes this.

stefanhaller commented on pkg/gui/controllers/helpers/commits_helper.go:
```diff
@@ -19,10 +19,9 @@ type ICommitsHelper interface {
-	// getVerifyFlag                 func() bool
+	getCommitSummary              func() string
+	setCommitSummary              func(string)
+	getCommitDescription          func() string
```
This should go into the previous commit, I suppose. Also below.

kschweiger replied:
True. I kind of expected you would squash on merge 😄

stefanhaller replied:
No, we don't do that here. On the contrary, I value clean and well-structured commit histories very much. (Not just for the end result of a branch when it is merged, but also during review, so I prefer a [fixup-based workflow](https://github.com/jesseduffield/lazygit/blob/master/docs/Fixup_Commits.md) over just piling commits onto one another, which is what we are now doing here. Something to keep in mind for next time, it's too late for this one. 😄)

stefanhaller commented on pkg/integration/tests/commit/commit_skip_hooks.go:
```diff
@@ -0,0 +1,35 @@
+
+		t.Views().Commits().Focus()
+		t.Views().Main().Content(Contains("foo bar"))
+		t.Views().Extras().Content(Contains("--no-verify"))
```
This is an interesting way of checking whether we passed the no-verify flag, but I'm not sure I like it much. If you had a test that committed twice, once with and once without the flag, you couldn't use this check for the second commit.

I'd prefer it if we installed an actual hook (using the technique in `fail_hooks_then_commit_no_hooks.go`), and have it write something to a file; we can then check whether the file exists (or what it contains) using `t.FileSystem().PathPresent()` or `t.FileSystem().FileContent()`.

stefanhaller reviewed (commented):
This is a great start, but it does still need quite a bit of work; see comments below.

kschweiger commented on pkg/commands/git_commands/commit.go:
```diff
@@ -85,13 +85,11 @@ func (self *CommitCommands) ResetToCommit(hash string, strength string, envVars
-
 	cmdArgs := NewGitCmd("commit").
-		ArgIf(skipHookPrefix != "" && strings.HasPrefix(summary, skipHookPrefix), "--no-verify").
+		ArgIf(verify, "--no-verify").
```
I adapted the code as you suggested with [97e05e5](https://github.com/jesseduffield/lazygit/pull/4374/commits/97e05e5febcf61b298c60ab156ccf7beb1fd2a6f)

kschweiger reviewed (commented):
I adapted this test to use an actual hook but  I'm not quite certain this is what you meant. So this is more a draft and I'm happy to reiterate and update the other integration (e.g. [CommitSwitchToEditorSkipHooks](https://github.com/jesseduffield/lazygit/blob/2a4911c68f450d33377bda8e7c0471bf28394fee/pkg/integration/tests/commit/commit_switch_to_editor_skip_hooks.go)) tests if necessary.

Like this, the hook would update the file content and if you replace `keys.Files.CommitChangesWithoutHook` with `keys.Files.CommitChanges` the test fails.  

Overall I'm fine with it and it is certainly more explicit but I'm not sure it is that much clearer than just looking at the `--no-verify` in the command log for this particular test.

stefanhaller commented:
> I adapted this test to use an actual hook but I'm not quite certain this is what you meant. 

Yes, this is close to what I meant, but I was wrong. 😄 We can simplify the tests quite a bit by installing a hook that always fails. This way we don't need to check for the existence of files, or for text appearing in the Extras view, in order to tell whether the hook ran; we can tell whether it ran or not simply by whether the commit succeeded. I pushed 98ac34074b18 to do that for some of the tests; this needs further adaptions for the others, and I think we now have some redundancy between the tests. I was too lazy to look at that in more detail, can you please clean this up?

Another thing I noticed now (and missed in the first round of review) is that there's unnecessary work happening with storing the forceSkipHooks flag in the OpenCommitMessagePanelOpts, and then passing it back to the OnConfirm handler. This is unnecessary and confusing because it is only used by the OnConfirm of the HandleCommitPressWithMessage call; all other callsites ignored the flag, and the concept of skipping hooks doesn't even make sense e.g. for creating a tag. It's much cleaner to simply capture the flag in the OnConfirm lambda in the only place where we need it; I pushed 2a869799ef96 to do that.

Feel free to squash everything into one commit now, and then to push further fixups for the remaining test adaptations.

kschweiger commented:
I squashed all the previous commit and refactored the tests. 

Only `fail_hooks_then_commit_no_hooks.go` explicitly tests that the hook fail with the regular commit. I'm usually a fan of making integration tests rather explicit even if not really necessary.  So I added a shared function for the other tests that basically tries to commit w/ hooks, fails and resets the working directory. In this case my rational is that they all share the same "hook" and if somebody down the line would decide to change it, the tests might still succeed even though the hook would not fail.

stefanhaller commented:
@kschweiger Great work, looks good to me. I agree that it's good to add the sanity check that the hook works at the beginning of each test.

I had to add a fixup (47bec2d75fec) because one of the tests failed after rebasing onto #4346. Will merge when green, thanks for the contribution!

stefanhaller commented:
Oh, and I also added a commit to remove the obsolete text.
